### PR TITLE
Revert "Revert "Docs, part 8""

### PR DESCRIPTION
**Pull request by arogachev**
Reverts yiisoft/validator#457. See description to #445,

### CODE REVIEW

File: src/DataSet/ObjectDataSet.php [hunk 1]
```diff
@@ -9,26 +9,170 @@
...
+ *
+ * Note that the rule attributes can be combined with others without affecting parsing. Which properties to parse can be
+ * configured via {@see ObjectDataSet::$propertyVisibility} and {@see ObjectDataSet::$skipStaticProperties} options.
+ *
+ * The other combinations of rules / data are also possible, for example: the data is provided by implementing
+ * {@see DataSetInterface} and rules are parsed from the attributes.
+ *
+ * Please refer to the guide for more examples.
+ *
+ * Rules and data provided via separate methods have the highest priority over attributes and properties so, when using
```
Comment (samdark) [inline src/DataSet/ObjectDataSet.php:135]:
```suggestion
 * Rules and data provided via separate methods have the highest priority over attributes and properties so, when used
```

File: src/DataSet/ObjectDataSet.php [hunk 2]
```diff
@@ -9,26 +9,170 @@
...
+ * together, the latter ones will be ignored without exception.
+ *
+ * When {@see RuleInterface} / {@see DataSetInterface} are not implemented, uses {@see ObjectParser} and supports
+ * caching for data and attribute methods (partially) and rules (completely) which can be disabled on demand.
+ *
+ * For getting only rules by a class string, use {@see AttributesRulesProvider} instead.
```
Comment (samdark) [inline src/DataSet/ObjectDataSet.php:141]:
```suggestion
 * For getting only rules by a class name string, use {@see AttributesRulesProvider} instead.
```

File: src/Helper/ObjectParser.php [hunk 3]
```diff
@@ -19,35 +19,150 @@
...
+ *
+ * ```php
+ * [
+ *     'title' => 'Yii3 Overview 3',
+ *     'author' => 'John',
+ *     'files' => [],
+ * ];
+ * ```
+ *
+ * A class string is valid as a source too. This way only rules will be parsed:
```
Comment (samdark) [inline src/Helper/ObjectParser.php:97]:
```suggestion
 * A class name string is valid as a source too. This way only rules will be parsed:
```

**Review by samdark — APPROVED** [review]

File: src/DataSet/ObjectDataSet.php [hunk 4]
```diff
@@ -9,26 +9,170 @@
...
 use Yiisoft\Validator\RulesProviderInterface;
 
 /**
- * This data set makes use of attributes introduced in PHP 8. It simplifies rules configuration process, especially for
- * nested data and relations. Please refer to the guide for examples.
+ * A data set for object data. The object passed to this data set can provide rules and data by implementing
+ * {@see RuleInterface} and {@see DataSetInterface}. Alternatively this data set allows getting rules from PHP
+ * attributes (attached to class properties and class itself) and data from object properties.
+ *
+ * An example of object implementing {@see RuleInterface}:
```
Comment (vjik) [inline src/DataSet/ObjectDataSet.php:21]:
```suggestion
 * An example of object implementing {@see RulesProviderInterface}:
```

File: src/DataSet/ObjectDataSet.php [hunk 5]
```diff
@@ -9,26 +9,170 @@
...
 use Yiisoft\Validator\Helper\ObjectParser;
+use Yiisoft\Validator\RuleInterface;
+use Yiisoft\Validator\RulesProvider\AttributesRulesProvider;
 use Yiisoft\Validator\RulesProviderInterface;
 
 /**
- * This data set makes use of attributes introduced in PHP 8. It simplifies rules configuration process, especially for
- * nested data and relations. Please refer to the guide for examples.
+ * A data set for object data. The object passed to this data set can provide rules and data by implementing
+ * {@see RuleInterface} and {@see DataSetInterface}. Alternatively this data set allows getting rules from PHP
```
Comment (vjik) [inline src/DataSet/ObjectDataSet.php:18]:
```suggestion
 * {@see RulesProviderInterface} and {@see DataSetInterface}. Alternatively this data set allows getting rules from PHP
```

File: src/DataSet/ObjectDataSet.php [hunk 6]
```diff
@@ -9,26 +9,170 @@
...
+ *
+ * The other combinations of rules / data are also possible, for example: the data is provided by implementing
+ * {@see DataSetInterface} and rules are parsed from the attributes.
+ *
+ * Please refer to the guide for more examples.
+ *
+ * Rules and data provided via separate methods have the highest priority over attributes and properties so, when using
+ * together, the latter ones will be ignored without exception.
+ *
+ * When {@see RuleInterface} / {@see DataSetInterface} are not implemented, uses {@see ObjectParser} and supports
```
Comment (vjik) [inline src/DataSet/ObjectDataSet.php:138]:
```suggestion
 * When {@see RulesProviderInterface} / {@see DataSetInterface} are not implemented, uses {@see ObjectParser} and supports
```

File: src/DataSet/ObjectDataSet.php [hunk 7]
```diff
@@ -90,6 +273,12 @@ public function getData(): mixed
         return $this->parser->getData();
     }
 
+    /**
+     * An optional attribute names translator. It's taken from the {@see $object} when
+     * {@see AttributeTranslatorProviderInterface} is implemented. In case of it's missing, a `null` value is returned.
+     *
+     * @return AttributeTranslatorInterface|null An attribute translator instance or `null if it was not provided.
```
Comment (vjik) [inline src/DataSet/ObjectDataSet.php:280]:
```suggestion
     * @return AttributeTranslatorInterface|null An attribute translator instance or `null` if it was not provided.
```

File: src/Helper/ObjectParser.php [hunk 8]
```diff
@@ -136,20 +294,20 @@ public function getAttributeTranslator(): ?AttributeTranslatorInterface
...
 
     /**
-     * @return array<string, ReflectionProperty>
+     * Returns Reflection properties parsed from {@see $source} in accordance with {@see $propertyVisibility} and
+     * {@see $skipStaticProperties} values. Repetitive calls utilize cache if it's enabled in {@see $useCache}.
+     *
+     * @return array<string, ReflectionProperty> A mapping between Reflection property names and their values.
      */
-    public function getReflectionProperties(): array
+    private function getReflectionProperties(): array
```
Comment (vjik) [inline src/Helper/ObjectParser.php:302]: Keep this method as public. It's need for `FormCollector` implement in `yiisoft/form`.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Helper/ObjectParser.php (hunk 3 above)
Comment (arogachev) [reply src/Helper/ObjectParser.php]: Fixed along with other occurrences.

**Review by vjik — APPROVED** [review]

**Pull request merged**